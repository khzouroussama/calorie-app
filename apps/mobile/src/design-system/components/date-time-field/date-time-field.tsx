import { format } from 'date-fns';

import { TextField, TextFieldProps } from '../text-field';
import { Box } from '../box';

import DateTimePickerModal from 'react-native-modal-datetime-picker';
import { Pressable } from '../pressable';
import React, { forwardRef, memo } from 'react';

export type DateTimeFieldProps = {
  min?: Date;
  max?: Date;
  mode?: 'date' | 'time' | 'datetime';
  label?: string;
  onChange?: (date: Date) => void;
  value: Date;
  error?: string | undefined;
  placeholder?: string;
};

export const DateTimeField = memo(
  forwardRef<unknown, DateTimeFieldProps>(function FormDateTimeField(props) {
    const { value, error, onChange, ...inputProps } = props;

    const [isDatePickerVisible, setDatePickerVisibility] =
      React.useState(false);

    const showDatePicker = () => {
      setDatePickerVisibility(true);
    };

    const hideDatePicker = () => {
      setDatePickerVisibility(false);
    };

    const handleConfirm = (date: Date) => {
      onChange(date);
      hideDatePicker();
    };

    return (
      <Box>
        <Pressable onPress={showDatePicker}>
          <TextField
            editable={false}
            {...inputProps}
            error={error}
            value={
              !value
                ? undefined
                : format(
                    value,
                    (() => {
                      switch (props.mode) {
                        case 'date':
                          return 'yyyy-MM-dd';
                        case 'time':
                          return 'HH:mm';
                        case 'datetime':
                          return 'yyyy-MM-dd HH:mm';
                      }
                    })(),
                  )
            }
          />
        </Pressable>

        <DateTimePickerModal
          isVisible={isDatePickerVisible}
          mode={props.mode || 'date'}
          date={value ?? new Date()}
          onConfirm={handleConfirm}
          onCancel={hideDatePicker}
          maximumDate={props.max}
          minimumDate={props.min}
        />
      </Box>
    );
  }),
);
