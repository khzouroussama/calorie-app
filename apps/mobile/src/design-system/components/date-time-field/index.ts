export { DateTimeField, DateTimeFieldProps } from './date-time-field';
