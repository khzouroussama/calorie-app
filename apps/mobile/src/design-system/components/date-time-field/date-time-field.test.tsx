import React from 'react';
import renderer from 'react-test-renderer';

import { DateTimeField } from './date-time-field';

Date.now = jest.fn(() => new Date(Date.UTC(2023, 1, 14)).valueOf());

describe('<Card />', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<DateTimeField value={new Date(Date.now())} mode="datetime" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
