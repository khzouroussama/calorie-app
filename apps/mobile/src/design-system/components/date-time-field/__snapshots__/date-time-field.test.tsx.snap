// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> renders correctly 1`] = `
<View
  style={
    [
      undefined,
      undefined,
    ]
  }
>
  <RNGestureHandlerButton
    collapsable={false}
    delayLongPress={600}
    enabled={true}
    exclusive={true}
    handlerTag={1}
    handlerType="NativeViewGestureHandler"
    onGestureEvent={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onHandlerStateChange={[Function]}
    rippleColor={0}
    touchSoundDisabled={false}
  >
    <View
      accessible={true}
      collapsable={false}
      style={{}}
    >
      <View
        animatedStyle={
          {
            "value": {
              "transform": [
                {
                  "scale": 1,
                },
              ],
            },
          }
        }
        collapsable={false}
        style={
          {
            "transform": [
              {
                "scale": 1,
              },
            ],
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "#F4F2F1",
                  "borderColor": "#B6ACA6",
                  "borderRadius": 4,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "overflow": "hidden",
                },
                undefined,
              ]
            }
          >
            <TextInput
              editable={false}
              mode="datetime"
              placeholderTextColor="#564E4A"
              style={
                {
                  "alignSelf": "stretch",
                  "color": "#191015",
                  "flex": 1,
                  "fontFamily": "spaceGroteskRegular",
                  "fontSize": 16,
                  "height": 24,
                  "marginHorizontal": 12,
                  "marginVertical": 8,
                  "paddingHorizontal": 0,
                  "paddingVertical": 0,
                }
              }
              textAlignVertical="top"
              underlineColorAndroid="rgba(0, 0, 0, 0)"
              value="2023-02-14 01:00"
            />
          </View>
        </View>
      </View>
    </View>
  </RNGestureHandlerButton>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    isVisible={false}
    onHide={[Function]}
    transparent={true}
    visible={false}
  />
</View>
`;
